core/trino-spi/src/main/java/io/trino/spi/type/Decimals.java

```diff
@@ -297,15 +297,28 @@ public static long rescale(long value, int fromScale, int toScale)
         if (toScale < fromScale) {
             throw new IllegalArgumentException("target scale must be larger than source scale");
         }
-        return value * longTenToNth(toScale - fromScale);
+
+        long returnValue = value * longTenToNth(toScale - fromScale);
+
+        if (overflows(returnValue, MAX_SHORT_PRECISION)) {
```

Comment:
nit: this could have perf impact potentially (minimal, but still) in cases where reader knows that all values will fit.

Reply:
Should we make the overflow check explicit too in the separate function and leave it up to the user of the code to decide whether they want to check for overflow or not?